### ISSUE TITLE
Net header issue when using musl libc

### ISSUE THREAD

**Issue by newbluemoon** [stsp/dosemu2#530]
Error message:

```
In file included from libpacket.c:26:0:
 /usr/include/netinet/if_ether.h:101:8: error: redefinition of 'struct ethhdr'
  struct ethhdr {
         ^~~~~~
 In file included from /usr/include/linux/if_tun.h:20:0,
                  from libpacket.c:25:
 /usr/include/linux/if_ether.h:141:8: note: originally defined here
  struct ethhdr {
         ^~~~~~

```

https://wiki.musl-libc.org/faq.html:

> Q: Why am I getting “error: redefinition of struct ethhdr/tcphdr/etc”?
> The application you’re trying to compile mixes userspace and kernel headers
> (/include/linux). The kernel headers are notoriously broken for userspace and
> clash with the definitions provided by musl. It only works (in certain cases)
> with GLIBC because…
> 
> - the userspace headers just include the kernel headers, and
> - because the kernel headers have a hardcoded compatibility fix for GLIBC to
>   prevent exactly this issue (linux/libc_compat.h)
> 
> the issue can be fixed in different ways:
> 
> - remove the linux/ include directives
> - use the patched kernel-headers package from sabotage linux
> - factor out the code needing the kernel headers into a separate compilation
>   unit which doesn’t use any userspace headers
> - copy the required struct definitions into an internal header and rename them.
> 

src/dosext/net/libpacket.c includes:
-  net/ethernet.h
-  netinet/if_ether.h
-  linux/if_tun.h

glibc include chain:
  net/ethernet.h --> linux/if_ether.h --> struct ethhdr
  netinet/if_ether.h --> linux/if_ether.h
  linux/if_tun.h --> linux/if_ether.h

musl include chain:
  net/ethernet.h --> netinet/if_ether.h --> struct ethhdr
  linux/if_tun.h --> linux/if_ether.h --> struct ethhdr

Compilation doesn’t throw an error with following workaround(?):
```
--- src/dosext/net/libpacket.c.orig	2017-11-08 15:26:02.000000000 +0100
+++ src/dosext/net/libpacket.c	2017-11-13 19:44:05.964811045 +0100
@@ -23,10 +23,12 @@
 #include <net/if.h>
 #include <netinet/in.h>
 #include <linux/if_tun.h>
-#include <netinet/if_ether.h>
 #include <netpacket/packet.h>
-#include <net/ethernet.h>
 #include <assert.h>
+#ifdef __GLIBC__
+# include <net/ethernet.h>
+# include <netinet/if_ether.h>
+#endif
 
 #include "emu.h"
 #include "utilities.h"

```

Note: There is no `__MUSL__` macro. See https://wiki.musl-libc.org/faq.html:

> The answer is no, it won't be added, because it's a bug to assume a
> certain implementation has particular properties rather than testing.
> So far, every time somebody's asked for this with a particular usage
> case in mind, the usage case was badly wrong, and would have broken
> support for the next release of musl...

So I don’t think this patch would be generally applicable (if it is correct).

Removing `linux/if_tun.h` leads to
```
libpacket.c: In function 'tun_alloc':
libpacket.c:269:23: error: 'IFF_TAP' undeclared (first use in this function); did you mean 'IFF_UP'?
       ifr.ifr_flags = IFF_TAP | IFF_NO_PI;
                       ^~~~~~~
                       IFF_UP
libpacket.c:269:23: note: each undeclared identifier is reported only once for each function it appears in
libpacket.c:269:33: error: 'IFF_NO_PI' undeclared (first use in this function); did you mean 'IFF_NOARP'?
       ifr.ifr_flags = IFF_TAP | IFF_NO_PI;
                                 ^~~~~~~~~
                                 IFF_NOARP
libpacket.c:274:23: error: 'TUNSETIFF' undeclared (first use in this function); did you mean 'TCSETAF'?
       err = ioctl(fd, TUNSETIFF, (void *) &ifr);
                       ^~~~~~~~~
                       TCSETAF
```

**Comment by stsp** [comment]
Should now be fixed.

**Comment by newbluemoon** [comment]
Thanks! You’re fantastic. The build finished now :)

I just forgot about one thing (sorry!): there is also no `netipx/ipx.h`, but `linux/ipx.h` so I had the following patch applied automatically:

```
--- src/dosext/net/ipx.c.orig	2017-11-08 15:26:02.000000000 +0100
+++ src/dosext/net/ipx.c	2017-11-13 15:03:14.040917938 +0100
@@ -23,9 +23,15 @@
 #include <sys/socket.h>
 #include <sys/time.h>
 #include <sys/types.h>
-#include <netipx/ipx.h>
 #include <netinet/in.h>
 #include <errno.h>
+#ifndef __GLIBC__
+# include <linux/ipx.h>
+# define IPX_TYPE 1
+# define SOL_IPX 256
+#else
+# include <netipx/ipx.h>
+#endif
 
 #include "emu.h"
 #include "timers.h"
--- src/dosext/net/ipxglt.c.orig	2017-11-08 15:26:02.000000000 +0100
+++ src/dosext/net/ipxglt.c	2017-11-13 15:06:48.056952388 +0100
@@ -22,9 +22,15 @@
 #include <sys/time.h>
 #include <sys/ioctl.h>
 #include <net/route.h>
-#include <netipx/ipx.h>
 #include <netinet/in.h>
 #include <errno.h>
+#ifndef __GLIBC__
+# include <linux/ipx.h>
+# define IPX_TYPE 1
+# define SOL_IPX 256
+#else
+# include <netipx/ipx.h>
+#endif
 
 #include "emu.h"
 #include "utilities.h"

```

But I don’t think that’s good to do in general. I think it’s better to have a check for `netipx/ipx.h` and if that doesn’t exist check for `linux/ipx.h`. Should I open another issue?

**Comment by stsp** [comment]
Should now be fixed.

**Comment by newbluemoon** [comment]
Perfect! Thanks again!

**Comment by stsp** [comment]
So does the whole thing work?

**Comment by newbluemoon** [comment]
I’ll do some actual testing tomorrow and will let you know about the outcome :-)

**Comment by newbluemoon** [comment]
So far everything seems to be fine. :) Everything I ever did with dosemu works: running a text processor, a database program and some games. Sound works. I just haven’t ever done networking with dosemu (which would be the most interesting part, I guess), so I need to check it out a bit more. For now I have set up a tap device which I can ping. Will do some further testing.

**Comment by stsp** [comment]
Cool! Were there the protected mode/DPMI
programs among the ones you tested?
Could you please list the plugins there were
built? (`ls *.so` in the `bin` dir of the build tree,
and also the list is printed by configure).

> set up a tap device which I can ping. Will do some further testing.

You can as well set up the user-space networking
that does not involve TAP or any configuration
whatsoever. For this you need to build vde2-devel
with the patches that dosemu2 has in
`src/plugin/vde/patches`.
Then re-build dosemu2 and make sure it built the
vde plugin.
Yes, I know, not the best practice for the end-user,
but this is all I could offer so far. TAP requires root
privs and a bit of a set-up (although quite well
documented in dosemu docs).
Unfortunately I don't have time to code up the
network stack in a slirp-alike fashion, but that
would be very good:
https://github.com/stsp/dosemu2/issues/167#issuecomment-223811277

**Comment by newbluemoon** [comment]
configure log:
```
Enabling plugins:
        libao
        gpm
        alsa
        ladspa
        extra_charsets
        term
        X
        sdl
        midimisc
        fluidsynth
        munt
        charsets
        console
        modemu

```

```
%  ls -1 *.so
libplugin_alsa.so
libplugin_console.so
libplugin_fluidsynth.so
libplugin_gpm.so
libplugin_ladspa.so
libplugin_libao.so
libplugin_munt.so
libplugin_sdl.so
libplugin_term.so
libplugin_X.so
```

Configure lists more than there are .so’s. Are some bundled in one .so?

DPMI is not working. When starting dosemu I get:

    ERROR: DPMI is not supported on your kernel ( >= 3.14 ), sorry!

But with the glibc build as well. I wanted to ask about this anyway ;) Wasn’t able to figure this out and since I didn’t need it ignored it. I set `$_cpu_emu = "full"`. Tried to run the ’links’ browser, have tried both csdpmi5b.zip and csdpmi7b.zip, but got:

    DPMI is not supported on your linux kernel!
    Protected mode not accessible.
    Load error: no DPMI – Get csdpmi*b.zip

I’m using Void Linux. Could it be that a kernel security feature causes this? Or are there some settings I’m not aware of?

Next I’ll test the vde plugin.

**Comment by stsp** [comment]
What kernel version?
`uname -r`

> Configure lists more than there are .so’s. Are some bundled in one .so?

Some are bundled inside the main binary -
the compile-time plugins.

**Comment by newbluemoon** [comment]
```
% uname -r
4.9.59_1
```

**Comment by stsp** [comment]
Look for `CONFIG_MODIFY_LDT_SYSCALL`
in `/boot/config-<your-kernel>`

**Comment by newbluemoon** [comment]
CONFIG_MODIFY_LDT_SYSCALL=y

**Comment by stsp** [comment]
`CONFIG_X86_ESPFIX64`?

**Comment by newbluemoon** [comment]
CONFIG_X86_ESPFIX64

This one doesn’t exist.

**Comment by stsp** [comment]
`CONFIG_X86_16BIT`?

**Comment by newbluemoon** [comment]
# CONFIG_X86_16BIT is not set

**Comment by stsp** [comment]
Should now be "fixed".

**Comment by newbluemoon** [comment]
Thanks! Is there any drawback when using those options? Else I would open an issue in voidlinux/void-packages asking for them.

**Comment by stsp** [comment]
I am not aware of any security problems
these days. In 3.14 the 16bit support was
considered a security problem and disabled,
but 3.16 introduced `CONFIG_X86_ESPFIX64`
that was supposed to close the security hole,
so they were re-enabled by default.
Your distro altered the default options.

**Comment by newbluemoon** [comment]
Ok. I’ll try my luck then ;) and again thanks for all your efforts!

**Comment by stsp** [comment]
This also means your DOS test-suit is quite
small. You can try windows-3.1 or D/L some
protected mode game like duke3d or carmageddon
to test DPMI.

**Comment by newbluemoon** [comment]
Will try. As mentioned earlier it’s sufficient for my needs, but of course it’d be good to have all features available!

**Comment by newbluemoon** [comment]
So I got an answer regarding the kernel options: “These were removed as part of hardening the x86_64 kernel.“ So I’ll build my own kernel and will also test it on i686.

**Comment by stsp** [comment]
Does void linux manifest some security
features? From the overview on their site,
seems not:
```
Void is a general purpose operating system, based on the monolithic Linux® kernel. Its package
system allows you to quickly install, update and remove software; software is provided in binary
packages or can be built directly from sources with the help of the XBPS source packages collection.
```

So I wonder what kind of hardening they mean.
Note that this also severely breaks wine (win16
components won't work).

Your manual testing will help, thanks!
I think home users should stay away from
security-hardening distros to not have to
unharden them the way you now have to.

**Comment by newbluemoon** [comment]
> So I wonder what kind of hardening they mean.

Regarding the kernel there’s [this commit](https://github.com/voidlinux/void-packages/commit/7ef2c2705f3b16566fcbe06ac3e10d506979b7b1).
Compilers use `--enable-default-pie` and `--enable-default-ssp`, added flags are `-fstack-protector-strong -D_FORTIFY_SOURCE=2`.

I’m not qualified to assess this in any way, but I really do like the distro.

**Comment by stsp** [comment]
I can see in this patch also
```
-CONFIG_MODIFY_LDT_SYSCALL=y
+# CONFIG_MODIFY_LDT_SYSCALL is not set
```

and you say it was `y`.
So it makes a bit more sense to disable
this during a hardening process, rather
than leave it enabled but disable the
supplementary options. So it would be
nice to find out why you have that option
still enabled, and if so - others should be
enabled too, otherwise its just broken for
no reason.

I think general-purpose distros should not
withdraw the features from the user, like
the ability to run dosemu and wine, and
backing that with the security hardening
argument.

**Comment by newbluemoon** [comment]
> I can see in this patch also
> 
> -CONFIG_MODIFY_LDT_SYSCALL=y
> +# CONFIG_MODIFY_LDT_SYSCALL is not set
> and you say it was y.

It was changed back [here](https://github.com/voidlinux/void-packages/commit/79fdce105d4c74f088c3327b24a0244584a21046).

**Comment by stsp** [comment]
Ok, so they revert changes for "go"
but refuse to do so for dosemu2.
I won't install that distro. :)

**Comment by stsp** [comment]
> Maybe the maintainers will think about it when it’s being reviewed. ;-)

So it will be visible to the maintainers that
this doesn't work yet, and needs "this and
that" to get working? Then that would be fine.
OTOH if you get the dosemu2 package into
the distro that does not allow DPMI and it
will hit the end-users in that form, then this
will be quite bad because users will attribute
the blame to dosemu2 itself (dosbox works,
lets use dosbox).

**Comment by newbluemoon** [comment]
I would add a message that is displayed when installing the package making the user aware of this problem.

Or do you suggest I shouldn’t do it? I won’t do it if you don’t want it.

**Comment by stsp** [comment]
I am not sure as I have never seen this
distro and how easy is it to "fix the problem"
by compiling a new kernel, how visible is this
warning, etc etc.
If it is so easy, why wouldn't it be possible
for the package's rules to check this capability
and, if needed, build the new kernel automatically?
Or maybe depend on the kernel from some
kind of "PPA" as in ubuntu?
Of course I have no idea what am I talking about. :)
Have you fixed this problem locally and do the
things work for you fine now?

**Comment by newbluemoon** [comment]
> how easy is it to "fix the problem"

Ok, maybe not that easy for new users, but I think the majority of the user is of the more experienced kind. I guess it also depends on how eager one is to learn new things.

> how visible is this warning

The package manager is in text mode, though there is a GUI (which I never used). The message is shown at the end of the installation process.

> Or maybe depend on the kernel from some
kind of "PPA" as in ubuntu?

It would be possible to have another kernel with those options in the repository, but as far as I can tell this is not wanted. I think there was a discussion about adding kernels with different features, but don’t remember exactly where. I do remember that the arguments against having all sorts of different kernels made sense to me. Sorry I can’t be more specific.

> Have you fixed this problem locally and do the
things work for you fine now?

I have built the modified kernel but couldn’t do a reboot, yet. Hopefully later tonight (doing too many things at once ;-)).

**Comment by stsp** [comment]
Yes, having an extra kernels is unwanted simply
because it is a security threat: they won't be all
up-to-date with the security fixes.
But automatically building the kernel may be more
interesting, if possible at all... I guess not an
option unless you are gentoo that builds everything
on fly.

**Comment by newbluemoon** [comment]
> But automatically building the kernel may be more
interesting, if possible at all... I guess not an
option

You guess right :)

**Comment by newbluemoon** [comment]
With the modified kernel dosemu2 can make use of DPMI now. Test program was duke3d. Worked flawlessly on glibc, but immediately segfaults when starting on musl. Non-DPMI programs work without problems, though. I just don’t know how to debug this (as ironically some debug code was disabled to make it run in the first place).

**Comment by stsp** [comment]
The disabled debug code was supplementary.
Please compile with debug info
(`./default-configure -d`), open another ticket
and attach the `-D9+M` log to it.
I don't hope that porting DPMI to new libc can
be done from just a log, but we can try.
